apache/superset
# fix(chart): deprecate persisting url_params

### SUMMARY
Currently there are a few problems in Explore view:
- if `url_params` are present in chart metadata, it's impossible to set new URL params to a chart in Explore view
- When saving a chart, the following error toast is emitted: "Form data not found in cache, reverting to chart metadata.". See screenshot below:
![image](https://user-images.githubusercontent.com/33317356/155984186-b9ef5b9a-63c8-4910-8fe6-d5004a4df4bb.png)

This PR changes how the hidden control [`url_params`](https://github.com/apache/superset/blob/2072225a8637299c23c337966c9d5aaaeaaa732b/superset-frontend/src/explore/controlPanels/sections.tsx#L56-L64) is handled by no longer be persisting it in chart metadata or cached chart form data. This is done by changing the following:
- Remove the error message about form data missing from cache if `form_data_key` is not present in the URL (sort of unrelated, but a regression nonetheless).
- Remove the `url_params` property from `form_data` if present before saving chart metadata
- Adds the removed `url_params` back to the URL during saving in the frontend so that the reloaded Explore/Dashboard view will retain the URL params that were present before saving chart metadata
- Adds an integration test for `CreateChartCommand` (bycatch - it was missing) and updates related unit tests

Note that charts that currently rely on default URL params that are persisted in chart data will still work like before - however, when the chart is saved/overwritten, the `url_params` state will be lost. So when making changes to old charts that are using the `url_param` function in templating code and haven't set default value as the second argument, users will need to update the `url_param` call to include the default value.

### TESTING INSTRUCTIONS
1. enable `ENABLE_TEMPLATE_PROCESSING` feature flag
4. Crete a new chart using the `flights` dataset
5. Add a groupby for `AIRLINE`, a metric for `COUNT(*)` and a filter for `AIRLINE IN ('{{ url_param('test') }}')`
6. Add the param `test=AA` to the URL
7. Save chart
8. See error toast (no longer happens after PR)
9. remove the URL param and reload the chart
10. Notice that the URL param `test=AA` reappears (no longer happens after PR)

### ADDITIONAL INFORMATION
